mindmeld/models/nn_utils/layers.py

```diff
@@ -0,0 +1,269 @@
+                    self.embeddings.weight.data[idx] = torch.as_tensor(emb)
+                msg = f"Initialized {len(embedding_weights)} number of embedding weights " \
+                      f"from the embedder model"
+                logger.info(msg)
+            else:
+                # when weights are passed as an array or tensor
+                self.embeddings.load_state_dict({'weight': torch.as_tensor(embedding_weights)})
+        self.embeddings.weight.requires_grad = update_embeddings
+
+        self.embedding_for_coefficients = None
+        if coefficients is not None:
+            if not len(coefficients) == num_tokens:
+                msg = f"Length of coefficients ({len(coefficients)}) must match the number of " \
+                      f"embeddings ({num_tokens})"
+                raise ValueError(msg)
+            self.embedding_for_coefficients = nn.Embedding(num_tokens, 1, padding_idx=padding_idx)
+            self.embedding_for_coefficients.load_state_dict({'weight': coefficients})
+            self.embedding_for_coefficients.weight.requires_grad = update_coefficients
+
+        self.dropout = nn.Dropout(embeddings_dropout)
+
+    def forward(self, padded_token_ids):
+        # padded_token_ids: dim: [BS, SEQ_LEN]
+        # returns:          dim: [BS, SEQ_LEN, EMB_DIM]
+
+        # [BS, SEQ_LEN] -> [BS, SEQ_LEN, EMB_DIM]
+        outputs = self.embeddings(padded_token_ids)
+
+        if self.embedding_for_coefficients:
+            # [BS, SEQ_LEN] -> [BS, SEQ_LEN, 1]
+            coefficients = self.embedding_for_coefficients(padded_token_ids)
+            # [BS, SEQ_LEN, EMB_DIM] -> [BS, SEQ_LEN, EMB_DIM]
+            outputs = torch.mul(outputs, coefficients)
+
+        outputs = self.dropout(outputs)
+
+        return outputs
+
+
+class CnnLayer(nn_module):
+
+    def __init__(self, emb_dim, kernel_sizes, num_kernels):
+        super().__init__()
+
+        if isinstance(num_kernels, list) and len(num_kernels) != len(kernel_sizes):
+            # incorrect length of num_kernels list specified
+            num_kernels = [num_kernels[0]] * len(kernel_sizes)
+        elif isinstance(num_kernels, int) and num_kernels > 0:
+            # num_kernels is a single integer value
+            num_kernels = [num_kernels] * len(kernel_sizes)
+        elif not isinstance(num_kernels, list):
+            msg = f"Invalid value for num_kernels: {num_kernels}. " \
+                  f"Expected a list of same length as emb_dim ({len(emb_dim)})"
+            raise ValueError(msg)
+
+        self.convs = nn.ModuleList()
+        # Unsqueeze input dim [BS, SEQ_LEN, EMD_DIM] to [BS, 1, SEQ_LEN, EMDDIM] and send as input
+        # Each conv module output's dimensions are [BS, n, SEQ_LEN, 1]
+        for length, n in zip(kernel_sizes, num_kernels):
+            self.convs.append(
+                nn.Sequential(
+                    nn.Conv2d(1, n, (length, emb_dim), padding=(length - 1, 0),
```

Comment:
what does `padding=(length - 1, 0)` mean? The torch does say its the implicit padding on each side, I am not entirely sure what that means here